File: frontend/src/data/graph.ts
import type { SceneNode, GraphEdge } from '../types';

export const sampleNodes: SceneNode[] = [
  { id: 'iris', x: 320, y: 200, label: 'Iris', initials: 'IR', type: 'hub' },
  { id: 'wren', x: 180, y: 120, label: 'Wren', initials: 'WR' },
  { id: 'cole', x: 460, y: 120, label: 'Cole', initials: 'CO' },
  { id: 'jon', x: 180, y: 300, label: 'Jon', initials: 'JN' },
  { id: 'mara', x: 460, y: 300, label: 'Mara', initials: 'MA' },
  { id: 'kai', x: 120, y: 210, label: 'Kai', initials: 'KA' },
  { id: 'fen', x: 520, y: 210, label: 'Fen', initials: 'FN', type: 'minor' },
  { id: 'doc', x: 320, y: 380, label: 'Doc', initials: 'DC', type: 'minor' },
  { id: 'sib', x: 400, y: 380, label: 'Sib', initials: 'SB', type: 'minor' },
];

export const sampleEdges: GraphEdge[] = [
  { a: 'iris', b: 'wren', kind: 'family', weight: 2 },
  { a: 'iris', b: 'cole', kind: 'conflict', weight: 3 },
  { a: 'iris', b: 'jon', kind: 'romance', weight: 2 },
  { a: 'iris', b: 'mara', kind: 'family', weight: 2 },
  { a: 'iris', b: 'kai', kind: 'mentor', weight: 2 },
  { a: 'wren', b: 'cole', kind: 'conflict', weight: 1 },
  { a: 'wren', b: 'jon', kind: 'alliance', weight: 1 },
  { a: 'cole', b: 'mara', kind: 'secret', weight: 2 },
  { a: 'cole', b: 'fen', kind: 'alliance', weight: 1 },
  { a: 'jon', b: 'doc', kind: 'alliance', weight: 1 },
  { a: 'mara', b: 'sib', kind: 'family', weight: 1 },
  { a: 'fen', b: 'doc', kind: 'secret', weight: 1 },
];

File: backend/app/models/graph.py
import enum
import uuid

from sqlalchemy import Float, ForeignKey, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Mapped, mapped_column

from app.models.base import Base, OrgScopedMixin


class NodeType(str, enum.Enum):
    hub = "hub"
    minor = "minor"


class EdgeKind(str, enum.Enum):
    conflict = "conflict"
    alliance = "alliance"
    romance = "romance"
    mentor = "mentor"
    secret = "secret"
    family = "family"


class EdgeProvenance(str, enum.Enum):
    author = "author"
    ai_accepted = "ai_accepted"
    ai_pending = "ai_pending"
    scaffold = "scaffold"


class CharacterNode(Base, OrgScopedMixin):
    __tablename__ = "character_nodes"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    story_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("stories.id"))
    character_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("characters.id"))
    x: Mapped[float] = mapped_column(Float, default=0.0)
    y: Mapped[float] = mapped_column(Float, default=0.0)
    label: Mapped[str] = mapped_column(String(255))
    initials: Mapped[str] = mapped_column(String(10))
    node_type: Mapped[NodeType | None] = mapped_column(nullable=True)
    first_appearance_scene: Mapped[int] = mapped_column(default=1)


class CharacterEdge(Base, OrgScopedMixin):
    __tablename__ = "character_edges"
    __table_args__ = (
        UniqueConstraint("story_id", "source_node_id", "target_node_id", name="uq_edge"),
    )

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    story_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("stories.id"))
    source_node_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("character_nodes.id"))
    target_node_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("character_nodes.id"))
    kind: Mapped[EdgeKind] = mapped_column()
    weight: Mapped[float] = mapped_column(Float, default=0.5)
    provenance: Mapped[EdgeProvenance] = mapped_column(default=EdgeProvenance.author)
    evidence: Mapped[list] = mapped_column(JSONB, default=list)
    first_evidenced_scene: Mapped[int] = mapped_column(default=1)
